### PR TITLE
Remove child aware callbacks when parent lifecycle is destroyed

### PR DESCRIPTION
**Pull request by CherryPerry**
The following code will:

1. Produce memory leak of `view`, because it is captured by callback, which is not removed later.
2. Be invoked twice for old and new instance of `view`.

```kotlin
override fun onViewCreated(view: View, lifecycle: Lifecycle) {
  whenChildAttached { child, commonLifecycle ->
    commonLifecycle.createDestroy { bind(view to child) }
  }
}
```

PR adds destroyed state checks (API usage correctness) and callback removal when parent lifecycle is destroyed.

TODO: Remove default lifecycle

### CODE REVIEW

**Comment by CherryPerry** [comment]
Try to use `Thread.stackTrace` to verify lifecycle (check `onViewCreated` method call).

**Comment by CherryPerry** [comment]
Screenshot 2021-05-25 at 14 21 00

File: sandbox/src/main/java/com/badoo/ribs/sandbox/rib/switcher/SwitcherInteractor.kt [hunk 1]
```diff
@@ -97,13 +98,15 @@ internal class SwitcherInteractor(
     override fun onCreate(nodeLifecycle: Lifecycle) {
         super.onCreate(nodeLifecycle)
 
-        createDestroy<Blocker> { blocker ->
-            bind(blocker.output to blockerOutputConsumer)
-        }
+        childAware(nodeLifecycle) {
```
Comment (CherryPerry) [inline sandbox/src/main/java/com/badoo/ribs/sandbox/rib/switcher/SwitcherInteractor.kt:101]: The main part.

**Comment by zsoltk** [comment]
Parts from the screenshot as new PR? Here it's enough to fix immediate issue, right?

File: libraries/rib-base/src/main/java/com/badoo/ribs/clienthelper/childaware/ChildAwareDSL.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package com.badoo.ribs.clienthelper.childaware
+
+import androidx.lifecycle.Lifecycle
+import com.badoo.ribs.core.Rib
+
+class ChildAwareDSL(
```
Comment (zsoltk) [inline libraries/rib-base/src/main/java/com/badoo/ribs/clienthelper/childaware/ChildAwareDSL.kt:6]:
```suggestion
class ChildAwareScope(
```

Comment (CherryPerry) [reply libraries/rib-base/src/main/java/com/badoo/ribs/clienthelper/childaware/ChildAwareDSL.kt:6]: Done

**Review by zsoltk — APPROVED** [review]

**Pull request merged**